File: leetcode/00079_word_search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        if not word:
            return True
        rows = len(board)
        if not rows:
            return False
        cols = len(board[0])
        dx_dy = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        used = set()
        def solve(row: int, col: int, widx: int) -> bool:
            if (
                row < 0 or row >= rows or
                col < 0 or col >= cols or
                board[row][col] != word[widx]
            ):
                return False
            if widx + 1 == len(word):
                return True
            for dx, dy in dx_dy:
                nrow = row + dx
                ncol = col + dy
                if (nrow, ncol) not in used:
                    used.add((nrow, ncol))
                    if solve(nrow, ncol, widx + 1):
                        return True
                    used.remove((nrow, ncol))
        for i in range(rows):
            for j in range(cols):
                used = {(i, j)}
                if solve(i, j, 0):
                    return True
        return False

File: leetcode/00079_word_search.rs
impl Solution {
    pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
        fn solve(board: &mut [Vec<char>], w: &[char], r: usize, c: usize) -> bool {
            if board[r][c] != w[0] { return false; }
            if w.len() == 1 { return true; }
            board[r][c] = '*';
            if r > 0 && solve(board, &w[1..], r - 1, c) { return true; }
            if c > 0 && solve(board, &w[1..], r, c - 1) { return true; }
            if r + 1 < board.len() && solve(board, &w[1..], r + 1, c) { return true; }
            if c + 1 < board[0].len() && solve(board, &w[1..], r, c + 1) { return true; }
            board[r][c] = w[0];
            false
        }

        let mut board = board;
        let word = word.chars().collect::<Vec<char>>();
        
        for row in 0..board.len() {
            for col in 0..board[0].len() {
                if solve(&mut board, &word, row, col) {
                    return true;
                }
            }
        }
        false
    }
}

